common/flatpak-dir.c

```diff
@@ -7502,6 +7505,149 @@ flatpak_dir_create_system_child_repo (FlatpakDir   *self,
   return flatpak_dir_create_child_repo (self, cache_dir, file_lock, optional_commit, error);
 }
 
+/* Convert an #EpcAppFilterOarsValue to an #AsContentRatingValue. This is
+ * actually a trivial cast, since the types are defined the same; but throw in
+ * a static assertion to be sure. */
+static AsContentRatingValue
+convert_app_filter_oars_value (EpcAppFilterOarsValue filter_value)
+{
+  G_STATIC_ASSERT (AS_CONTENT_RATING_VALUE_LAST == EPC_APP_FILTER_OARS_VALUE_INTENSE + 1);
+
+  return (AsContentRatingValue) filter_value;
+}
+
+/* Get the AppStream data for @ref, loading it from the locally cached AppStream
+ * XML. If the AppStream file doesn’t exist, or if @ref_string is not listed in
+ * it, %NULL will be returned without an error. An error will be returned if
+ * parsing the AppStream file fails.
+ *
+ * If @arch is %NULL, the default architecture will be used.
+ *
+ * If the @ref_string is for a runtime, %NULL will be returned immediately with
+ * no error.
+ *
+ * A strong reference is returned on the #AsApp.
+ */
+static AsApp *
+get_appstream_for_ref (FlatpakDir          *self,
+                       FlatpakRemoteState  *state,
+                       const char          *ref_string,
+                       const char          *arch,
+                       GCancellable        *cancellable,
+                       GError             **error)
+{
+  g_autoptr(FlatpakRef) ref = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  ref = flatpak_ref_parse (ref_string, error);
+  if (ref == NULL)
+    return NULL;
+
+  /* Runtimes don’t have appstream. */
+  if (flatpak_ref_get_kind (ref) != FLATPAK_REF_KIND_APP)
+    return NULL;
+
+  if (arch == NULL)
+    arch = flatpak_get_arch ();
+
+  g_autofree char *install_path = g_file_get_path (flatpak_dir_get_path (self));
+  g_autoptr(AsStore) store = as_store_new ();
+  g_autofree char *appstream_path = NULL;
+  g_autofree char *unique_id = NULL;
+
+  if (flatpak_dir_get_remote_oci (self, state->remote_name))
+    appstream_path = g_build_filename (install_path, "appstream", state->remote_name,
+                                       arch, "appstream.xml.gz",
+                                       NULL);
+  else
+    appstream_path = g_build_filename (install_path, "appstream", state->remote_name,
+                                       arch, "active", "appstream.xml.gz",
+                                       NULL);
+
+  g_autoptr(GFile) appstream_file = g_file_new_for_path (appstream_path);
+  as_store_from_file (store, appstream_file, NULL, cancellable, &local_error);
```

Comment:
I have no idea; I can’t find much information or code about it. Based on https://blogs.gnome.org/hughsie/2018/09/20/speeding-up-appstream-mmaping-xml-using-libxmlb/, it looks like gnome-software has been ported (see the `xb_` calls in some of its plugins), and the binary XML library is an alternative to libappstream-glib, rather than a modification of it. In any case, this is not something we need to worry about until flatpak is ported to it; and if we share the flatpak code for loading appstream data (as per your comment below), we should get it for free (unless we need to port away from libappstream-glib). Sigh.